ANGLE: Simplify PoolAllocator further

kkinnunen-apple opened the pull request:
#### 51961a5a88419fefa2fb98de35fbd979e2c72bcc
<pre>
ANGLE: Simplify PoolAllocator further
<a href="https://bugs.webkit.org/show_bug.cgi?id=297041">https://bugs.webkit.org/show_bug.cgi?id=297041</a>
<a href="https://rdar.apple.com/157725945">rdar://157725945</a>

Reviewed by Dan Glastonbury.

- Remove distinct implementations of the member functions when
  ANGLE_DISABLE_POOL_ALLOC is defined
- Support guards with ANGLE_DISABLE_POOL_ALLOC.
- Hardcode alignment
- Store the segments in a std::vector instead linked list entries inside
  the pools
- Avoid having padding guard and after guard, just have padding guard
  if needed, otherwise after guard
- Increase pool size from 16k to 32k

* Source/ThirdParty/ANGLE/src/common/PoolAlloc.cpp:
(angle::PoolAllocator::Segment::Segment):
(angle::PoolAllocator::Segment::~Segment):
(angle::PoolAllocator::Segment::operator=):
(angle::PoolAllocator::Segment::Allocate):
(angle::PoolAllocator::Segment::data const):
(angle::PoolAllocator::~PoolAllocator):
(angle::PoolAllocator::addGuard):
(angle::PoolAllocator::reset):
(angle::PoolAllocator::allocateNewPoolSegment):
(angle::PoolAllocator::allocateSingleObject):
(angle::PoolAllocator::PoolAllocator): Deleted.
(angle::PoolAllocator::PageHeader::PageHeader): Deleted.
(angle::PoolAllocator::adjustAllocationExtent const): Deleted.
(angle::PoolAllocator::allocateNewPage): Deleted.
(angle::PoolAllocator::allocate): Deleted.
* Source/ThirdParty/ANGLE/src/common/PoolAlloc.h:
(angle::PoolAllocator::allocate):
(angle::PoolAllocator::adjustAllocationExtent const): Deleted.
(angle::PoolAllocator::addGuard): Deleted.
(angle::PoolAllocator::bump): Deleted.
* Source/ThirdParty/ANGLE/src/common/PoolAlloc_unittest.cpp:
(angle::TEST_F(PoolAllocatorTest, Interface)):
(angle::TEST_F(PoolAllocatorTest, Alignment)):
(angle::TEST_F(PoolAllocatorTest, ResetRecyclesMemory)):
(angle::TEST_F(PoolAllocatorGuardTest, AlignmentGuardDetectsOverflowWrite)):
(angle::TEST_F(PoolAllocatorGuardTest, AllocationGuardsDetectsOverflowWrite)):
(angle::TEST_F(PoolAllocatorGuardTest, AllocationGuardsDetectsUnderflowWrite)):
(angle::PoolAllocatorTest::GetAlignment): Deleted.
(angle::TEST_P): Deleted.
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/vulkan/AllocatorHelperPool.h:

Canonical link: <a href="https://commits.webkit.org/298409@main">https://commits.webkit.org/298409@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79be5b67c1e518f7003623833e974ee025eb1579

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115397 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35091 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25593 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121486 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65975 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d01cae6c-586a-42a5-be99-5e065a21d023) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35750 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43662 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87681 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b49e8337-d7a1-49dd-8b70-305849082322) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118345 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28513 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103586 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68073 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27671 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21708 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65143 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97901 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21821 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124651 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42336 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31715 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96462 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42704 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99774 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96248 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24490 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41476 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19332 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38246 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42217 "Built successfully") | | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41720 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45048 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43439 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/vulkan/AllocatorHelperPool.h:
```diff
@@ -56,8 +56,8 @@ class DedicatedCommandBlockPool final
     static_assert(sizeof(CommandHeaderIDType) < kCommandHeaderSize, "Check size of CommandHeader");
-    // Pool Alloc uses 16kB pages w/ 16byte header = 16368bytes. To minimize waste
-    //  using a 16368/12 = 1364. Also better perf than 1024 due to fewer block allocations
+    // PoolAllocator uses 16384 byte pools. To minimize waste using a 16384/12 = 1365.
```
The comment and constant don't match. Is this intentional?

djg reviewed (approved):
:shipit:

kkinnunen-apple committed:
ANGLE: Simplify PoolAllocator further
https://bugs.webkit.org/show_bug.cgi?id=297041
rdar://157725945

Reviewed by Dan Glastonbury.

- Remove distinct implementations of the member functions when
  ANGLE_DISABLE_POOL_ALLOC is defined
- Support guards with ANGLE_DISABLE_POOL_ALLOC.
- Hardcode alignment
- Store the segments in a std::vector instead linked list entries inside
  the pools
- Avoid having padding guard and after guard, just have padding guard
  if needed, otherwise after guard
- Increase pool size from 16k to 32k

* Source/ThirdParty/ANGLE/src/common/PoolAlloc.cpp:
(angle::PoolAllocator::Segment::Segment):
(angle::PoolAllocator::Segment::~Segment):
(angle::PoolAllocator::Segment::operator=):
(angle::PoolAllocator::Segment::Allocate):
(angle::PoolAllocator::Segment::data const):
(angle::PoolAllocator::~PoolAllocator):
(angle::PoolAllocator::addGuard):
(angle::PoolAllocator::reset):
(angle::PoolAllocator::allocateNewPoolSegment):
(angle::PoolAllocator::allocateSingleObject):
(angle::PoolAllocator::PoolAllocator): Deleted.
(angle::PoolAllocator::PageHeader::PageHeader): Deleted.
(angle::PoolAllocator::adjustAllocationExtent const): Deleted.
(angle::PoolAllocator::allocateNewPage): Deleted.
(angle::PoolAllocator::allocate): Deleted.
* Source/ThirdParty/ANGLE/src/common/PoolAlloc.h:
(angle::PoolAllocator::allocate):
(angle::PoolAllocator::adjustAllocationExtent const): Deleted.
(angle::PoolAllocator::addGuard): Deleted.
(angle::PoolAllocator::bump): Deleted.
* Source/ThirdParty/ANGLE/src/common/PoolAlloc_unittest.cpp:
(angle::TEST_F(PoolAllocatorTest, Interface)):
(angle::TEST_F(PoolAllocatorTest, Alignment)):
(angle::TEST_F(PoolAllocatorTest, ResetRecyclesMemory)):
(angle::TEST_F(PoolAllocatorGuardTest, AlignmentGuardDetectsOverflowWrite)):
(angle::TEST_F(PoolAllocatorGuardTest, AllocationGuardsDetectsOverflowWrite)):
(angle::TEST_F(PoolAllocatorGuardTest, AllocationGuardsDetectsUnderflowWrite)):
(angle::PoolAllocatorTest::GetAlignment): Deleted.
(angle::TEST_P): Deleted.
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/vulkan/AllocatorHelperPool.h:

Canonical link: https://commits.webkit.org/298409@main
